File: 0051-n-queens/0051-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        board = [["."]*n for _ in range(n)]
        res = []

        self.helper(0, res, board)

        return res
    
    def helper(self, row, res, board) -> None:
        # Base case
        if (row == len(board)):
            res.append([''.join(row) for row in board])
            return
        
        # Recursive case
        for col in range(0, len(board)):
            if (self.isValid(row, col, board)):
                board[row][col] = "Q"
                self.helper(row+1, res, board)
                board[row][col] = "."
    
    def isValid(self, row, col, board) -> bool:
        # Check column (towards UP)
        i = row
        while (i >= 0):
            if (board[i][col] == "Q"):
                return False
            i -= 1
        
        # Check Diagonal Left
        i = row
        j = col
        while (i >= 0 and j >= 0):
            if (board[i][j] == "Q"):
                return False
            i -= 1
            j -= 1

        # Check Digaonal Right
        i = row
        j = col
        while (i >= 0 and j < len(board)):
            if (board[i][j] == "Q"):
                return False
            i -= 1
            j += 1
        
        return True

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public boolean isSafe(int col,int row, int n,char[][] board){
        int duprow = row;
        int dupcol  = col;

        while(row>=0 && col>=0){
            if(board[row][col] ==  'Q') return false;
            row--;
            col--;
        }

        row = duprow;
        col = dupcol;

        while(col>=0){
            if(board[row][col] == 'Q') return false;
            col--;
        }

        col = dupcol;

        while(row<n && col>=0){
            if(board[row][col] == 'Q') return false;
            row++;
            col--;
        }

        return true;
    }
    public List<String> construct(char[][] board){
        List<String> rowadd = new ArrayList<>();
        for(int i =0;i<board.length;i++){
            String row = new String(board[i]);
            rowadd.add(row);
        }
        return rowadd;
    }
    public void solve(List<List<String>> ans,char[][] board,int n ,int col){
        if(col == n){
            ans.add(construct(board));
            return;
        }

        for(int row = 0;row<n;row++){
            if(isSafe(col,row,n,board)){
                board[row][col] = 'Q';
                solve(ans,board,n,col+1);
                board[row][col] = '.';
            }
        }

    }
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        solve(ans,board,n,0);
        return ans;
    }
}